src/core/server/ui_settings/saved_objects/dynamic_config_controlled_ui_settings_wrapper.ts

```diff
@@ -0,0 +1,120 @@
+} from '../../../server';
+import { getWorkspaceState } from '../../../server/utils';
+import { isGlobalScope } from '../utils';
+import { InternalDynamicConfigServiceSetup } from '../../config';
+
+/**
+ * Wrapper for reading dynamic feature flag to decide if global UI settings
+ * are editable or not for non-admin users
+ * @param isPermissionControlEnabled
+ * @param dynamicConfig
+ */
+export class DynamicConfigControlledUiSettingsWrapper {
+  /**
+   * @param isPermissionControlEnabled
+   */
+  constructor(
+    private readonly isPermissionControlEnabled: boolean,
+    private readonly dynamicConfig: InternalDynamicConfigServiceSetup
+  ) {}
+
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const updateUiSettingsWithPermission = async <T = unknown>(
+      type: string,
+      id: string,
+      attributes: Partial<T>,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<SavedObjectsUpdateResponse<T>> => {
+      const hasPermission = await this.checkPermission(wrapperOptions);
+      if (type === 'config' && !hasPermission && isGlobalScope(id)) {
+        throw this.generatePermissionError();
+      }
+
+      return wrapperOptions.client.update<T>(type, id, attributes, options);
+    };
+
+    const bulkUpdateWithPermission = async <T = unknown>(
+      objects: Array<SavedObjectsBulkUpdateObject<T>>,
+      options?: SavedObjectsBulkUpdateOptions
+    ) => {
+      const hasPermission = await this.checkPermission(wrapperOptions);
+      objects.forEach((obj) => {
+        if (obj.type === 'config' && !hasPermission && isGlobalScope(obj.id)) {
+          throw this.generatePermissionError();
+        }
+      });
+      return wrapperOptions.client.bulkUpdate<T>(objects, options);
+    };
+
+    return {
+      ...wrapperOptions.client,
+      create: wrapperOptions.client.create,
+      bulkCreate: wrapperOptions.client.bulkCreate,
+      delete: wrapperOptions.client.delete,
+      update: updateUiSettingsWithPermission,
+      bulkUpdate: bulkUpdateWithPermission,
+      get: wrapperOptions.client.get,
+      checkConflicts: wrapperOptions.client.checkConflicts,
+      errors: wrapperOptions.client.errors,
+      addToNamespaces: wrapperOptions.client.addToNamespaces,
+      deleteFromNamespaces: wrapperOptions.client.deleteFromNamespaces,
+      find: wrapperOptions.client.find,
+      bulkGet: wrapperOptions.client.bulkGet,
+      deleteByWorkspace: wrapperOptions.client.deleteByWorkspace,
+    };
+  };
+
+  private async checkPermission(
+    wrapperOptions: SavedObjectsClientWrapperOptions
+  ): Promise<boolean> {
+    // If saved object permission is disabled, everyone should be treated as admin here
+
+    const dynamicConfigServiceStart = await this.dynamicConfig.getStartService();
+    const store = dynamicConfigServiceStart.getAsyncLocalStore();
+    const client = dynamicConfigServiceStart.getClient();
+
+    try {
+      const dynamicConfig = await client.getConfig(
```

Comment:
Why we need to check the feature flag again here? Isn't it already being registered in the capabilities? 🤔